### PR TITLE
Update vesting contract

### PR DESCRIPTION
**Pull request by larry0x**
We want to extend unlocking by 3 months. Currently the contract does not allow changing the unlocking schedule.

This PR adds a method to allow the contract owner to change the config.

### CODE REVIEW

**Comment by dancreee** [comment]
@piobab @larry0x can you decide what to do about the failing build due to drop in code coverage

File: contracts/vesting/src/contract.rs [hunk 1]
```diff
@@ -316,7 +295,30 @@ pub fn query_positions(
...
+// Migration
+//--------------------------------------------------------------------------------------------------
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _: Env, _: Empty) -> Result<Response> {
+    use cw_storage_plus::Item;
+
+    const LEGACY_OWNER: Item<Addr> = Item::new("owner");
+    const LEGACY_VEST_DENOM: &str = "umars";
+    const LEGACY_UNLOCK_SCHEDULE: Item<Schedule> = Item::new("unlock_schedule");
```
[the remark this answers is not in the document]
Comment (dancreee) [reply contracts/vesting/src/contract.rs:313]: ok, nvm I see that you've removed the `state::{Position, OWNER, POSITIONS, UNLOCK_SCHEDULE},` and that's why we need to re-specify

File: contracts/vesting/src/contract.rs [hunk 2]
```diff
@@ -57,50 +61,45 @@ pub fn execute(deps: DepsMut, env: Env, info: MessageInfo, msg: ExecuteMsg) -> S
...
 
+pub fn update_config(
+    deps: DepsMut,
+    info: MessageInfo,
+    new_cfg: Config<String>,
+) -> Result<Response> {
+    let cfg = CONFIG.load(deps.storage)?;
+
+    // only owner can update config
+    if info.sender != cfg.owner {
```
Comment (piobab) [inline contracts/vesting/src/contract.rs:75]: btw in cosmwasm there is new macro:

```
ensure_eq!(info.sender, cfg.owner, Error::NotOwner);
```

File: contracts/vesting/src/contract.rs [hunk 3]
```diff
@@ -316,7 +295,30 @@ pub fn query_positions(
...
+    const LEGACY_UNLOCK_SCHEDULE: Item<Schedule> = Item::new("unlock_schedule");
+
+    let cfg = Config {
+        owner: LEGACY_OWNER.load(deps.storage)?,
+        denom: LEGACY_VEST_DENOM.into(),
+        unlock_schedule: LEGACY_UNLOCK_SCHEDULE.load(deps.storage)?,
+    };
+
+    CONFIG.save(deps.storage, &cfg)?;
+
```
Comment (piobab) [inline contracts/vesting/src/contract.rs:322]: Add `set_new_version`

**Review by piobab — APPROVED** [review]

**Comment by piobab** [comment]
@larry0x @dancreee pushed one commit for migration update & tests

**Comment by piobab** [comment]
@larry0x if you have problem with migration "unsupported optcode" on mars hub downgrade rust-optimizer.

**Comment by larry0x** [comment]
> @larry0x @dancreee pushed one commit for migration update & tests. @larry0x feel free to accept it or not.

Very nice

**Pull request merged**